sbol2/componentdefinition.py

```diff
@@ -134,7 +129,46 @@ def __init__(self, uri=URIRef("example"),
         self.sequenceConstraints = OwnedObject(self, SBOL_SEQUENCE_CONSTRAINTS,
                                                SequenceConstraint,
                                                '0', '*', None)
-        self._hidden_properties = [SBOL_SEQUENCE]
+        # This caches a sequence object so that it can be added
+        # to a document later
+        self._sequence_cache: Union[Sequence, None] = None
+
+    @property
+    def sequence(self):
+        seqs = self.sequences
+        if not seqs:
+            return None
+        seq_uri = seqs[0]
+        if self._sequence_cache and self._sequence_cache.identity == seq_uri:
+            return self._sequence_cache
+        # Not in the cache, try to look up in the document
+        if self.doc:
+            try:
+                return self.doc.find(seq_uri)
+            except NotImplementedError:
+                raise
+
+    @sequence.setter
+    def sequence(self, sequence: Union[Sequence, None]):
+        if not sequence:
+            # TODO: How to remove an orphaned sequence from the document?
+            self.sequences = None
+            self._sequence_cache = None
+            return
+        if self.doc:
+            try:
+                self.doc.add(sequence)
+            except SBOLError as e:
+                if e.error_code() != SBOLErrorCode.DUPLICATE_URI_ERROR:
+                    raise
+        self._sequence_cache = sequence
+        self.sequences = [sequence.identity]
+
+    def added_to_document(self, doc):
```

Comment:
Maybe make this private?  (In the super as well)

Reply:
I don't want to make it private because it is called by `Document`. I think it should be part of the public API of `SBOLObject` instances. We can discuss it and see what we think.

Reply:
I'm thinking it should be private because a casual user will never need to call it in the context of a script.  So it clutters the public API.  Even an advanced user would only need it if they are defining custom class behaviors, which probably means only us.

Since we can't make it private, is there any other way to hide it?  If there's nothing straightforward, then we should just leave it as is.

Reply:
I'm convinced. I changed the method name to `_added_to_document`.

Reply:
Nice